### PR TITLE
Faster DFL module without Conv

### PR DESCRIPTION
**Pull request by Y-T-G**
**DFL Conv - TRT FP32**

Takes 42.9s on MSCOCO
```
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 775.6±976.8 MB/s, size: 135.9 KB)
val: Scanning /datasets/coco/labels/val2017.cache... 4952 images, 48 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 5000/5000 51.7Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5000/5000 116.6it/s 42.9s
                   all       5000      36335      0.649      0.498      0.547      0.391
Speed: 1.0ms preprocess, 1.7ms inference, 0.0ms loss, 1.0ms postprocess per image
```

**DFL Matmul - TRT FP32**
Takes 42.5s on MSCOCO

```
val: Fast image access ✅ (ping: 0.0±0.0 ms, read: 670.3±556.3 MB/s, size: 145.3 KB)
val: Scanning /datasets/coco/labels/val2017.cache... 4952 images, 48 backgrounds, 0 corrupt: 100% ━━━━━━━━━━━━ 5000/5000 55.0Mit/s 0.0s
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 5000/5000 117.6it/s 42.5s
                   all       5000      36335      0.655      0.496      0.547      0.391
Speed: 1.0ms preprocess, 1.7ms inference, 0.0ms loss, 1.0ms postprocess per image
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
🚀 Replace the DFL projection Conv2d with a lightweight buffer-based projection for faster inference and simpler state.

### 📊 Key Changes
- 🧠 Removed the fixed `nn.Conv2d` (frozen weights) used to project DFL distributions to scalar offsets
- ⚡ Added a registered buffer `proj` holding the projection vector (arange) and used a matmul-based forward path
- 🧳 Implemented `__setstate__` to migrate older pickled checkpoints that still contain `conv` but not `proj`
- 🔁 Updated `forward()` to reshape, apply `softmax` over the distribution bins, and project via `self.proj @ x`

### 🎯 Purpose & Impact
- ⏱️ Improves runtime efficiency by avoiding a convolution op for a fixed projection
- 🧩 Simplifies module state by storing the projection as a buffer (cleaner, non-trainable by design)
- 🛠️ Preserves backward compatibility with older serialized models via state migration
- 📦 Potentially reduces overhead across detection models using DFL (especially at scale)

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:
-ब ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
-ब ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
-ब ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
-ब ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
-б ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
-б ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
-б ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

This is an automated note 🤖—an Ultralytics engineer will review and assist shortly. For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/nn/modules/block.py [hunk 1]
```diff
@@ -68,16 +68,20 @@ def __init__(self, c1: int = 16):
...
-        self.conv = nn.Conv2d(c1, 1, 1, bias=False).requires_grad_(False)
-        x = torch.arange(c1, dtype=torch.float)
-        self.conv.weight.data[:] = nn.Parameter(x.view(1, c1, 1, 1))
+        self.register_buffer("proj", torch.arange(c1, dtype=torch.float32).view(1, 1, 1, -1))
         self.c1 = c1
 
+    def __setstate__(self, state):
+        """Migrate old pickled models that have conv but no proj."""
+        self.__dict__.update(state)
+        if "proj" not in self._buffers:
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/block.py:77]: ⚠️ **HIGH**: `__setstate__` assumes `self.conv` exists when `proj` is missing. If a state dict/pickle lacks both `proj` and `conv`, this will raise `AttributeError` at the next line. Guard on `hasattr(self, "conv")` (or similar) before accessing it.

**Suggested change:**
```suggestion
        if "proj" not in self._buffers and hasattr(self, "conv"):
```

File: ultralytics/nn/modules/block.py [hunk 2]
```diff
@@ -68,16 +68,20 @@ def __init__(self, c1: int = 16):
...
+            self._buffers["proj"] = self.conv.weight.data.view(1, 1, 1, -1)
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/block.py:78]: 💡 **MEDIUM**: Directly mutating `self._buffers` bypasses `register_buffer` and may leave `self.proj` attribute unset in some edge cases (and is generally more fragile for future refactors). Prefer `self.register_buffer(...)` here to ensure both `_buffers` and the attribute are correctly populated.

**Suggested change:**
```suggestion
            self.register_buffer("proj", self.conv.weight.data.view(1, 1, 1, -1))
```

File: ultralytics/nn/modules/block.py [hunk 3]
```diff
@@ -68,16 +68,20 @@ def __init__(self, c1: int = 16):
...
+
     def forward(self, x: torch.Tensor) -> torch.Tensor:
         """Apply the DFL module to input tensor and return transformed output."""
-        b, _, a = x.shape  # batch, channels, anchors
-        return self.conv(x.view(b, 4, self.c1, a).transpose(2, 1).softmax(1)).view(b, 4, a)
-        # return self.conv(x.view(b, self.c1, 4, a).softmax(1)).view(b, 4, a)
+        b, _, a = x.shape
+        x = x.view(b, 4, self.c1, a).softmax(2)
+        return (self.proj @ x).view(b, 4, a)
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/modules/block.py:84]: 💡 **MEDIUM**: `proj` is fixed to `float32` while `x` may be `float16/bfloat16` during inference. This can trigger dtype promotion (extra casts) and hurt performance on GPU/AMP. Consider casting `proj` to `x.dtype` in the matmul path.

**Suggested change:**
```suggestion
        return (self.proj.to(dtype=x.dtype) @ x).view(b, 4, a)
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice simplification: moving the fixed DFL projection from a frozen Conv2d to a buffer-based matmul reduces state and should be faster/cleaner. Main issues to address are robustness in `__setstate__` (don’t assume `conv` exists when `proj` is missing, and avoid mutating `_buffers` directly) and potential dtype promotion in `forward()` when `x` runs under AMP.

💬 Posted 3 inline comments
